src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+ *
+ * If use_usec is true, measure "closest" in terms of bin start usec.
+ * If it is false, use bin index distance only.
+ */
+void
+circpad_machine_remove_closest_token(circpad_machineinfo_t *mi,
+                                     circpad_delay_t target_bin_us,
+                                     int use_usec)
+{
+  int lower = circpad_machine_first_lower_index(mi, target_bin_us);
+  int higher = circpad_machine_first_higher_index(mi, target_bin_us);
+  int current = circpad_histogram_usec_to_bin(mi, target_bin_us);
+  circpad_delay_t lower_us;
+  circpad_delay_t higher_us;
+
+  tor_assert(lower <= current);
+  tor_assert(higher >= current);
+
+  if (higher == mi->histogram_len && lower == -1) {
+    // Bins are empty
+    return;
+  } else if (higher == mi->histogram_len) {
+    // Higher bins are empty
+    tor_assert(mi->histogram[lower]);
+    mi->histogram[lower]--;
+    return;
+  } else if (lower == -1) {
+    // Lower bins are empty
+    tor_assert(mi->histogram[higher]);
+    mi->histogram[higher]--;
+    return;
+  }
+
+  if (use_usec) {
+    /* Find the closest bin midpoint to the target */
+    lower_us = circpad_histogram_bin_to_usec(mi, lower)/2 +
+                circpad_histogram_bin_to_usec(mi, lower+1)/2;
+    higher_us = circpad_histogram_bin_to_usec(mi, higher)/2 +
+                circpad_histogram_bin_to_usec(mi, higher+1)/2;
+
+    if (target_bin_us < lower_us) {
+      // Lower bin is closer
+      tor_assert(mi->histogram[lower]);
+      mi->histogram[lower]--;
+      return;
+    } else if (target_bin_us > higher_us) {
+      // Higher bin is closer
+      tor_assert(mi->histogram[higher]);
+      mi->histogram[higher]--;
+      return;
+    } else if (target_bin_us - lower_us > higher_us - target_bin_us) {
+      // Higher bin is closer
+      tor_assert(mi->histogram[higher]);
+      mi->histogram[higher]--;
+      return;
+    } else {
+      // Lower bin is closer
+      tor_assert(mi->histogram[lower]);
+      mi->histogram[lower]--;
+      return;
+    }
+  } else {
+    if (current - lower > higher - current) {
+      // Higher bin is closer
+      tor_assert(mi->histogram[higher]);
+      mi->histogram[higher]--;
+      return;
+    } else {
+      // Lower bin is closer
+      tor_assert(mi->histogram[lower]);
+      mi->histogram[lower]--;
+      return;
+    }
+  }
+}
+
+/**
+ * Remove a token from the exact bin corresponding to the target.
+ *
+ * If it is empty, do nothing.
+ */
+static void
+circpad_machine_remove_exact(circpad_machineinfo_t *mi,
+                             circpad_delay_t target_bin_us)
+{
+  int bin = circpad_histogram_usec_to_bin(mi, target_bin_us);
+
+  if (mi->histogram[bin] > 0)
+    mi->histogram[bin]--;
+}
+
+/**
+ * Check our state's cell limit count and tokens.
+ *
+ * Returns 1 if either limits are hit and we decide to change states,
+ * otherwise returns 0.
+ */
+static circpad_decision_t
+circpad_check_token_supply(circpad_machineinfo_t *mi)
```

Comment:
This function is a predicate; it should have a name that describes _what_ it is checking.

Reply:
Addressed in e52653c14d44f87df4f8a256bcf69a2a5b6775b9.